lib/Doctrine/ODM/MongoDB/Aggregation/Stage/Merge.php

```diff
@@ -0,0 +1,148 @@
+ */
+class Merge extends Stage
+{
+    private DocumentManager $dm;
+
+    /**
+     * @var string|array
+     * @psalm-var OutputCollection
+     */
+    private $into;
+
+    /** @var list<string> */
+    private array $on = [];
+
+    /** @var array<string, string> */
+    private array $let = [];
+
+    /**
+     * @var string|array|Builder
+     * @psalm-var WhenMatchedParam
+     */
+    private $whenMatched;
+
+    private ?string $whenNotMatched = null;
+
+    public function __construct(Builder $builder, DocumentManager $documentManager)
+    {
+        parent::__construct($builder);
+
+        $this->dm = $documentManager;
+    }
+
+    public function getExpression(): array
+    {
+        $params = (object) [
+            'into' => $this->into,
+        ];
+
+        if ($this->on) {
+            $params->on = count($this->on) === 1 ? $this->on[0] : $this->on;
+        }
+
+        if ($this->let) {
+            $params->let = $this->let;
+        }
+
+        if ($this->whenMatched) {
+            $params->whenMatched = $this->whenMatched instanceof Builder
+                ? $this->whenMatched->getPipeline(false)
+                : $this->whenMatched;
+        }
+
+        if ($this->whenNotMatched) {
+            $params->whenNotMatched = $this->whenNotMatched;
+        }
+
+        return ['$merge' => $params];
+    }
+
+    /**
+     * @param string|array $collection
+     * @psalm-param OutputCollection $collection
+     */
+    public function into($collection): self
+    {
+        if (is_array($collection)) {
+            $this->into = $collection;
+
+            return $this;
+        }
+
+        try {
+            $class      = $this->dm->getClassMetadata($collection);
+            $this->into = $class->getCollection();
+        } catch (MappingException $e) {
+            $this->into = $collection;
+        }
+
+        return $this;
+    }
+
+    /**
+     * Optional. Specifies variables to use in the pipeline stages.
+     *
+     * Use the variable expressions to access the fields from
+     * the joined collection's documents that are input to the pipeline.
+     *
+     * @param array<string, string> $let
+     */
+    public function let(array $let): self
+    {
+        $this->let = $let;
+
+        return $this;
+    }
+
+    public function on(string ...$fields): self
+    {
+        $this->on = array_values($fields);
+
+        return $this;
+    }
+
+    /**
+     * @param string|array|Builder|Stage $whenMatched
+     * @psalm-param WhenMatchedParam $whenMatched
+     */
+    public function whenMatched($whenMatched): self
+    {
+        if ($whenMatched instanceof Stage) {
+            $this->whenMatched = $whenMatched->builder;
+        } else {
+            $this->whenMatched = $whenMatched;
+        }
+
+        if ($this->builder === $this->whenMatched) {
+            throw new InvalidArgumentException('Cannot use the same Builder instance for $merge whenMatched pipeline.');
```

Comment:
Ignore this is the trailing period is consistent with other exceptions. I'm assuming that single-sentence messages don't need it, though.

```suggestion
            throw new InvalidArgumentException('Cannot use the same Builder instance for $merge whenMatched pipeline');
```

Reply:
Some use it, some don't. I usually include trailing periods, but I'm fine with either once we've decided on a format for ODM.